File: 10/main.py
from z3 import Int, Solver, sat, Sum

def parse_input():
    with open("input.txt") as f:
        machines = []
        for line in f.read().split("\n")[:-1]:
            [_, *buttons, joltages] = line.split(" ")
            buttons = [ [ int(i) for i in 
                          button[1:-1].split(",")
                          ] for button in buttons ]
            joltages = [ int(i) for i in 
                          joltages[1:-1].split(",") ]
            machines.append((buttons, joltages))
    return machines

def solve_machine(buttons, joltages):
    symbols = []
    s = Solver()
    for i in range(len(buttons)):
        symbol = Int("k"+str(i))
        symbols.append(symbol)
        s.add(symbol >= 0)
        
    for i in range(len(joltages)):
        s.add(Sum(symbols[j] for j in range(len(buttons)) if i in buttons[j]) == joltages[i])

    minimum = None
    while s.check() == sat:
        model = s.model()
        solution = {sym: model[sym].as_long() for sym in symbols}
        total = sum(solution.values())        
        minimum = total
        s.add(Sum(*symbols) < minimum)
    return minimum
    

machines = parse_input()

print("Part 2:", sum(solve_machine(*machine) for machine in machines))

File: 10/main.rs
use std::cmp::min;

struct Machine {
		target: String,
		buttons: Vec<Vec<usize>>,
}

fn parse_input() -> Vec<Machine> {
		std::fs::read_to_string("input.txt").unwrap()
				.lines()
				.map(|line| {
						let [mut light_pattern, button_tuples @ .., _] =
								&line.split(' ').collect::<Vec<_>>()[..] else { todo!() };
						light_pattern = light_pattern.strip_prefix('[').unwrap();
						light_pattern = light_pattern.strip_suffix(']').unwrap();
						let target = light_pattern.to_string();
						let buttons = button_tuples.iter().map(|&(mut tuple)| {
								tuple = tuple.strip_prefix('(').unwrap();
								tuple = tuple.strip_suffix(')').unwrap();
								tuple.split(',').map(|x| x.parse::<usize>().unwrap()).collect()
						}).collect();
						Machine{target, buttons}
				})
				.collect()
}

fn solve_machine_1(partial: &mut Vec<usize>, machine: &Machine) -> Option<usize> {
		if partial.len() == machine.buttons.len() {
				let mut joltages: Vec<_> = (0..machine.target.len()).map(|_| 0).collect();
				for i in 0..partial.len() {
						for &joltage in &machine.buttons[i] {
								joltages[joltage] += partial[i]
						}
				}
				if (0..machine.target.len()).all(|i| {
						match machine.target.chars().nth(i) {
								Some('#') => joltages[i] % 2 == 1,
								Some('.') => joltages[i] % 2 == 0,
								_ => todo!()
						}
				}) {
						return Some(partial.iter().sum())
				} else {
						return None
				}
		}
		let mut sols = vec![];
		for i in 0..=1 {
				partial.push(i);
				if let Some(cost) = solve_machine_1(partial, machine) {
						sols.push(cost);
				}
				partial.pop();
		}
		sols.iter().reduce(min).copied()
}

fn part_1(machine: &Machine) -> usize {
		let mut presses = vec![];
		return solve_machine_1(&mut presses, machine).unwrap()
}

fn solve(machines: &[Machine], func: fn(&Machine) -> usize) -> usize {
		machines.iter().map(func).sum()
}

fn main() {
		let machines = parse_input();
  	println!("Part 1: {}", solve(&machines, part_1));
}
